File: backend/internal/common/utils/logger.utils.go
package utils

import (
	"log"
	"os"
)

const (
	INFO = iota
	DEBUG
	WARN
	ERROR
)

type Logger struct {
	level  int
	logger *log.Logger
}

func NewLogger(level int) *Logger {
	return &Logger{
		level:  level,
		logger: log.New(os.Stdout, "", log.LstdFlags),
	}
}

func (l *Logger) Debug(msg string, v ...interface{}) {
	if l.level <= DEBUG {
		l.logger.SetPrefix("DEBUG: ")
		l.logger.Printf(msg, v...)
	}
}

func (l *Logger) Info(msg string, v ...interface{}) {
	if l.level <= INFO {
		l.logger.SetPrefix("INFO: ")
		l.logger.Printf(msg, v...)
	}
}

func (l *Logger) Warn(msg string, v ...interface{}) {
	if l.level <= WARN {
		l.logger.SetPrefix("WARN: ")
		l.logger.Printf(msg, v...)
	}
}

func (l *Logger) Error(msg string, v ...interface{}) {
	if l.level <= ERROR {
		l.logger.SetPrefix("ERROR: ")
		l.logger.Printf(msg, v...)
	}
}

File: sandbox/javascript/src/utils/logger.utils.js
const {stdout} = require('node:process')

class Logger {
    static instance = new Logger([stdout]);

    constructor(streams = []) {
        this.streams = streams;
    }

    #writeToStreams(level, message, ...args) {
        message = this.#formatMessage(level, message, ...args)
        this.streams.forEach(stream => {
            stream.write(message);
        });
    }

    log(message, ...args) {
        this.#writeToStreams('LOG', message, ...args)
    }

    info(message, ...args) {
        this.#writeToStreams('INFO', message, ...args)
    }

    warn(message, ...args) {
        this.#writeToStreams('WARN', message, ...args)
    }

    error(message, ...args) {
        this.#writeToStreams('ERROR', message, ...args)
    }

    #safeJSONStringify(data) {
        if (!data) return ''
        try {
            return JSON.stringify(data, null, 2)
        } catch {
            return data.toString ? data.toString() : ''
        }
    }

    #formatMessage(level, message, ...args) {
        const timestamp = new Date().toISOString()
        let metaData = ''
        for (const arg of args) {
            if (arg instanceof Error) {
                metaData += `${arg.toString()}\n${arg.stack}`
                continue
            }
            metaData += `${this.#safeJSONStringify(arg)}\n`
        }
        return `[${timestamp}] [${level.toUpperCase()}]: ${message} ${metaData}\n`
    }
}

module.exports = Logger
